### PR TITLE
Fix #8132. Open folder in integrated terminal.

### LINKED ISSUE #8132: Change Explorer Open In New Command Prompt Context Menu to use Integrated Terminal
- VSCode Version: 1.2+
- OS Version: ALL

I would like this to launch the integrated experience to aid in discover-ability.

Steps to Reproduce:
1. Right click on a file
2. Should open in integrated terminal with the right folder context

### PR DESCRIPTION
**Pull request by rebornix**
This is the first touch on Open Folder in Integrated Terminal. Some decisions still need to be revisited

* Configuration Name. Now I call it  `terminal.terminalKind` to align with  https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/common/debugProtocol.d.ts#L198 . We have put settings for integrated terminal in `parts/terminal` folder and external ones in `parts/execution`. I'm not sure putting this one into `parts/explorer` and name it as `explorer.terminalKind` would be a good idea as then we have settings for terminal in three different places.
* I didn't touch `runInTerminal` yet as it's up to the caller whether to use external or integrated terminal. For `openInTerminal`, we make the decision where we bind the action to the explorer. For `runInTerminal`, we make such decision when invoking terminals, e.g., 
https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/electron-browser/terminalSupport.ts#L21

### CODE REVIEW

File: src/vs/workbench/parts/execution/electron-browser/terminal.contribution.ts [hunk 1]
```diff
@@ -99,7 +112,18 @@ export class OpenConsoleAction extends Action {
 			}
 		}
 
-		this.terminalService.openTerminal(pathToOpen);
+		const configuration = this.configurationService.getConfiguration<ITerminalConfiguration>();
```
Comment (Tyriar) [inline src/vs/workbench/parts/execution/electron-browser/terminal.contribution.ts]: I think we actually want to make this decision in `ExplorerViewerActionContributor`, `OpenConsoleAction` is an action on the external/native terminal right now. Currently having `terminal.terminaKind` as `"integrated"` would make ctrl+shift+c behave similar to ctrl+` by creating and showing a new terminal - we want both to be able to be used at the same time.

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/parts/execution/electron-browser/terminal.contribution.ts (hunk 1 above)
Comment (rebornix) [inline src/vs/workbench/parts/execution/electron-browser/terminal.contribution.ts]: Nice catch, addressed.

**Comment by rebornix** [comment]
Now we register Open in Terminal/Integrated Terminal in Files part and put corresponding configuration in Files part as well.

File: src/vs/workbench/parts/execution/browser/execution.contribution.ts [hunk 2]
```diff
@@ -99,15 +60,75 @@ export class OpenConsoleAction extends Action {
...
+		id: string,
+		label: string,
+		@ITerminalService private terminalService: ITerminalService,
+		@IWorkbenchEditorService editorService: IWorkbenchEditorService,
+		@IWorkspaceContextService contextService: IWorkspaceContextService,
+		@IHistoryService historyService: IHistoryService
+	) {
+		super(id, label, editorService, contextService, historyService);
+
+		this.order = 49; // Allow other actions to position before or after
```
Comment (Tyriar) [inline src/vs/workbench/parts/execution/browser/execution.contribution.ts]: Can be pulled into `AbstarctOpenInTerminalAction`?

**Review by Tyriar — CHANGES_REQUESTED** [review]

**Comment by Tyriar** [comment]
@bpasero please review the changes in fileActions and the new `explorer.openInTerminalKind` settings key. The original proposal introduced a `terminal.terminalKind` key which is awkward because it would represent a new settings category (or merging of external/integrated terminal categories):

image

Also because I really feel this belongs in the File Explorer category as the functionality is contained within that component:

image

File: src/vs/workbench/parts/execution/browser/execution.contribution.ts (hunk 2 above)
Comment (rebornix) [inline src/vs/workbench/parts/execution/browser/execution.contribution.ts]: Addressed.

**Comment by rebornix** [comment]
The context menu with new action

screen shot 2017-07-10 at 4 58 15 pm

File: src/vs/workbench/parts/files/browser/fileActions.contribution.ts [hunk 3]
```diff
@@ -21,6 +21,7 @@ import { FileStat, Model } from 'vs/workbench/parts/files/common/explorerModel';
 import { KeyMod, KeyChord, KeyCode } from 'vs/base/common/keyCodes';
 import { OpenFolderAction, OpenFileFolderAction, AddRootFolderAction, RemoveRootFolderAction } from 'vs/workbench/browser/actions/fileActions';
 import { copyFocusedFilesExplorerViewItem, revealInOSFocusedFilesExplorerItem, openFocusedExplorerItemSideBySideCommand, copyPathOfFocusedExplorerItem, copyPathCommand, revealInExplorerCommand, revealInOSCommand, openFolderPickerCommand, openWindowCommand, openFileInNewWindowCommand, deleteFocusedFilesExplorerViewItemCommand, moveFocusedFilesExplorerViewItemToTrashCommand, renameFocusedFilesExplorerViewItemCommand } from 'vs/workbench/parts/files/browser/fileCommands';
+import { ExplorerViewerActionContributor as TerminalExplorerViewerActionContributor } from 'vs/workbench/parts/execution/browser/terminal.contribution';
```
Comment (bpasero) [inline src/vs/workbench/parts/files/browser/fileActions.contribution.ts]: Compile error

File: src/vs/workbench/parts/files/browser/fileActions.contribution.ts [hunk 4]
```diff
@@ -185,6 +186,7 @@ class ExplorerViewersActionContributor extends ActionBarContributor {
 const actionBarRegistry = Registry.as<IActionBarRegistry>(ActionBarExtensions.Actionbar);
 actionBarRegistry.registerActionBarContributor(Scope.VIEWER, FilesViewerActionContributor);
 actionBarRegistry.registerActionBarContributor(Scope.VIEWER, ExplorerViewersActionContributor);
+actionBarRegistry.registerActionBarContributor(Scope.VIEWER, TerminalExplorerViewerActionContributor);
```
Comment (bpasero) [inline src/vs/workbench/parts/files/browser/fileActions.contribution.ts]: This should move into `execution-contribution.ts`

File: src/vs/workbench/parts/files/browser/files.contribution.ts [hunk 5]
```diff
@@ -334,6 +334,16 @@ configurationRegistry.registerConfiguration({
 				nls.localize('sortOrder.modified', 'Files and directories are sorted by last modified date, in descending order. Directories are displayed before files.')
 			],
 			'description': nls.localize('sortOrder', "Controls the way of sorting files and directories in the explorer.")
-		}
+		},
```
Comment (bpasero) [inline src/vs/workbench/parts/files/browser/files.contribution.ts]: This should be contributed from `execution.contribution.ts`

File: src/vs/workbench/parts/files/browser/files.contribution.ts [hunk 6]
```diff
@@ -334,6 +334,16 @@ configurationRegistry.registerConfiguration({
...
+		'explorer.openInTerminalKind': {
+			'type': 'string',
+			'enum': [
+				'external',
+				'integrated'
+			],
+			'description': nls.localize('explorer.openInTerminalKind', "Customizes what kind of terminal to launch."),
+			'default': 'external',
```
Comment (bpasero) [inline src/vs/workbench/parts/files/browser/files.contribution.ts]: Make the default "integrated"?

File: src/vs/workbench/parts/files/common/files.ts [hunk 7]
```diff
@@ -61,6 +61,7 @@ export interface IFilesConfiguration extends IFilesConfiguration, IWorkbenchEdit
 		autoReveal: boolean;
 		enableDragAndDrop: boolean;
 		sortOrder: SortOrder;
+		openInTerminalKind: string;
```
Comment (bpasero) [inline src/vs/workbench/parts/files/common/files.ts]: This is not the right place to have it, it should be in execution contribution.

**Review by bpasero — CHANGES_REQUESTED** [review]
@rebornix @Tyriar guys, we can do better :). The explorer code is now sprinkled with dependencies to `parts/execution`, does that make any sense? The way it used to be was correct: the terminal part would contribute into the explorer part and not force a dependency to the explorer. The explorer should only be concerned about managing and showing files. I have updated https://github.com/Microsoft/vscode/wiki/Code-Organization#workbench-parts

File: src/vs/workbench/parts/files/browser/files.contribution.ts (hunk 6 above)
Comment (Tyriar) [inline src/vs/workbench/parts/files/browser/files.contribution.ts]: :+1:

**Comment by Tyriar** [comment]
@rebornix let's go with something like `terminal.explorerKind` as a contribution within the Integrated Terminal settings category.

I created https://github.com/Microsoft/vscode/issues/30402 to do the eventual clean up I'd like to do.

**Comment by rebornix** [comment]
The final code change consists of 

* Option `terminal.explorerKind`
* Register Open In Integrated Terminal action when above option is set to `integrated` (which is the default value)
* Merge execution.contribution and terminal.contribution.
* Update test cases to avoid failure.

**Review by bpasero — APPROVED** [review]
Thanks, much better. Only noticed one nit: `AbstarctOpenInTerminalAction` (should be "abstract")

File: src/vs/workbench/parts/execution/electron-browser/execution.contribution.ts [hunk 8]
```diff
@@ -16,3 +36,192 @@ if (env.isWindows) {
...
+		this.terminalService.openTerminal(pathToOpen);
+
+		return TPromise.as(null);
+	}
+}
+
+export class OpenIntegratedTerminalAction extends AbstarctOpenInTerminalAction {
+
+	public static ID = 'workbench.action.terminal.openFolderInIntegratedTerminal';
+	public static Label = nls.localize('openFolderInIntegratedTerminal', "Open in Integrated Terminal");
```
Comment (Tyriar) [inline src/vs/workbench/parts/execution/electron-browser/execution.contribution.ts]: I think we should be using the same name here - adding "Integrated" makes the context menu quite a bit wider without adding much.

Comment (rebornix) [inline src/vs/workbench/parts/execution/electron-browser/execution.contribution.ts]: Revert to `Open in Terminal` now.

**Review by Tyriar — APPROVED** [review]

**Comment by Tyriar** [comment]
:fireworks:

**Pull request merged**